src/mysky/encrypted_files.ts

```diff
@@ -0,0 +1,330 @@
+  validateObject("fullData._data", _data, "parameter");
+  validateNumber("fullData._v", _v, "parameter");
+  validateUint8ArrayLen("key", key, "parameter", ENCRYPTION_KEY_LENGTH);
+
+  // Stringify the json and convert to bytes.
+  let data = stringToUint8ArrayUtf8(JSON.stringify(_data));
+
+  // Add padding so that the final size will be a padded block. The overhead will be added by encryption and we add the nonce at the end.
+  const totalOverhead = ENCRYPTION_OVERHEAD_LENGTH + ENCRYPTION_NONCE_LENGTH + ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH;
+  const finalSize = padFileSize(data.length + totalOverhead) - totalOverhead;
+  data = new Uint8Array([...data, ...new Uint8Array(finalSize - data.length)]);
+
+  // Generate a random nonce.
+  const nonce = new Uint8Array(randomBytes(ENCRYPTION_NONCE_LENGTH));
+
+  // Encrypt the data.
+  const encryptedBytes = secretbox(data, nonce, key);
+
+  // Prepend the metadata.
+  const metadata: EncryptedFileMetadata = { version: _v };
+  const metadataBytes = encodeEncryptedFileMetadata(metadata);
+  const finalBytes = new Uint8Array([...metadataBytes, ...encryptedBytes]);
+
+  // Prepend the nonce to the final data.
+  return new Uint8Array([...nonce, ...finalBytes]);
+}
+
+/**
+ * Derives key entropy for the given path seed.
+ *
+ * @param pathSeed - The given path seed.
+ * @returns - The key entropy.
+ */
+export function deriveEncryptedFileKeyEntropy(pathSeed: string): Uint8Array {
+  const bytes = new Uint8Array([...sha512("encryption"), ...sha512(pathSeed)]);
+  const hashBytes = sha512(bytes);
+  // Truncate the hash to the size of an encryption key.
+  return hashBytes.slice(0, ENCRYPTION_KEY_LENGTH);
+}
+
+/**
+ * Derives the encrypted file tweak for the given path seed.
+ *
+ * @param pathSeed - the given path seed.
+ * @returns - The encrypted file tweak.
+ */
+export function deriveEncryptedFileTweak(pathSeed: string): string {
+  let hashBytes = sha512(new Uint8Array([...sha512("encrypted filesystem tweak"), ...sha512(pathSeed)]));
+  // Truncate the hash or it will be rejected in skyd.
+  hashBytes = hashBytes.slice(0, HASH_LENGTH);
+  return toHexString(hashBytes);
+}
+
+/**
+ * Derives the path seed for the relative path, given the starting path seed and
+ * whether it is a directory. The path can be an absolute path if the root seed
+ * is given.
+ *
+ * @param pathSeed - The given starting path seed.
+ * @param subPath - The path.
+ * @param isDirectory - Whether the path is a directory.
+ * @returns - The path seed for the given path.
+ */
+export function deriveEncryptedFileSeed(pathSeed: string, subPath: string, isDirectory: boolean): string {
+  validateHexString("pathSeed", pathSeed, "parameter");
+  validateString("subPath", subPath, "parameter");
+  validateBoolean("isDirectory", isDirectory, "parameter");
+
+  let pathSeedBytes = hexToUint8Array(pathSeed);
+  subPath = sanitizePath(subPath);
+  const names = subPath.split("/");
+
+  for (const [index, name] of names.entries()) {
```

Comment:
use array.forEach instead of for-of for reason mentioned before

Reply:
```
Property 'forEach' does not exist on type 'IterableIterator<[number, string]>'. [2339]
```

Reply:
I think you made a mistake somewhere, `subPath` is a `string` so string split turns into `Array` - please try again and let me know if the issue persists (ex. `names.forEach((name, index) => { ... })`